### ISSUE TITLE
`OnlineSoftmaxWithSink::finalize` misses the exponent clamps its FA2 twin has — non-finite sink values yield NaN

### ISSUE THREAD

**Issue by headinthebox** [flashinfer-ai/flashinfer#4807]
## Description

`flashinfer/jit/attention/variants.py` contains two implementations of the sink softmax update. The FA2
one clamps its exponent differences; the FA3 one, 75 lines below in the same file, does not.

```cpp
// :39-43  — FA2  AttentionSink::update_m_d
float log_sink = (kv_tile_idx == 0 && qo_head_idx < params.num_qo_heads)
               ? params.sink[qo_head_idx] * math::log2e : -math::inf;
float m_new = (log_sink > m) ? log_sink : m;
scale       = math::ptx_exp2(max(m - m_new, -math::inf));                     // clamped
float d_new = math::ptx_exp2(max(log_sink - m_new, -math::inf)) + d * scale;  // clamped

// :117-125 — FA3  OnlineSoftmaxWithSink::finalize
float m_new = (log_sink > m) ? log_sink : m;
float scale = math::ptx_exp2(m - m_new);                                      // not clamped
float d_new = math::ptx_exp2(log_sink - m_new) + d * scale;                   // not clamped
scores_scale(mi) = pv_scale * scale / d_new;                                  // no d_new > 0 guard
```

PR #4401 added these clamps across the FA2 path, `state_t::merge`, `cascade.cuh`, `mla.cuh` and
`hopper/attention_updater.cuh` (including the `(sum > 0.f)` output-rescale guard at
`attention_updater.cuh:235`). The sink `update()` in this file was clamped in that PR;
`finalize()` was not.

## Reproduction of the arithmetic

Transcribing `finalize()` and evaluating it on the host (`clang -O0 -ffp-contract=off`, IEEE-754
binary32) — `pv_scale = 1`:

| `m` (row max) | `d` | `sink` | `scale` | `d_new` | `scores_scale` |
|---|---|---|---|---|---|
| `-inf` (fully masked row) | 0 | `-inf` | `nan` | `nan` | **NaN** |
| `-inf` | 0 | `-3.0` | 0 | 1 | ok |
| `-inf` | 0 | `-1e30` | 0 | 1 | ok |
| 2.0 | 5 | `-inf` | 1 | 5 | ok |
| **2.0** | **5** | **`+inf`** | 0 | `nan` | **NaN** |
| `-inf` | 0 | `+inf` | 0 | `nan` | **NaN** |
| 2.0 | 5 | `nan` | 1 | `nan` | **NaN** |
| 2.0 | 5 | `-3.0` | 1 | 5.03 | ok |

Two cases worth separating:

- **`sink = -inf` on a fully masked row** gives `(-inf) - (-inf)` in `m - m_new`. This is the same
  `(-inf) - (-inf)` shape #4401 fixed elsewhere, and it needs the row to be fully masked.
- **`sink = +inf`** gives `exp2(inf - inf)` in `d_new` and NaN's an *ordinary* row — no masked row
  involved.

A large-negative but finite sink is **fine** (`d_new = 1`), so the two clamps are the fix here; the
`d_new > 0` guard is defence in depth rather than the primary issue.

## Why these inputs can arrive

`sinks` is not validated anywhere in `flashinfer/prefill.py`, and is documented (`:2758`) as:

> Per-head attention-sink logits. When provided, the kernel applies the attention-with-sink variant: an
> additional virtual token whose logit is `sinks[head_idx]` is appended to each row of the softmax
> denominator. Shape `[num_qo_heads]`, dtype `float32`.

`-inf` is the conventional "contributes nothing" logit, and is a natural way to disable one head's sink
while keeping the tensor shape. The FA2 path constructs `-inf` for `log_sink` itself; the FA3
constructor (`:157`) is `log_sink = params.additional_params.sink[qo_head_idx] * math::log2e`, with no
such fallback, so a caller-supplied `-inf` reaches `finalize()` unchanged.

The existing tests generate sinks as `torch.rand(num_qo_heads, ...) * 5` (`tests/attention/test_attention_sink.py:238`,
`:417`) — finite and positive — so this region of the input space is not currently sampled.

## Proposed patch

Making `finalize()` match its twin above:

```diff
-      float scale = math::ptx_exp2(m - m_new);
-      float d_new = math::ptx_exp2(log_sink - m_new) + d * scale;
+      // Clamp as in AttentionSink::update_m_d above: max() drops the NaN from
+      // (-inf) - (-inf), and every difference is already <= 0 for finite inputs.
+      float scale = math::ptx_exp2(max(m - m_new, -math::inf));
+      float d_new = math::ptx_exp2(max(log_sink - m_new, -math::inf)) + d * scale;
...
-      scores_scale(mi) = pv_scale * scale / d_new;
+      scores_scale(mi) = (d_new > 0.f) ? pv_scale * scale / d_new : 0.f;
```

Since `m_new = max(m, log_sink)`, both differences are `<= 0` whenever the operands are finite, so the
clamps are inert on the finite inputs the tests cover.

## What I have and have not checked

- **Checked:** the source asymmetry above, on `main`; the arithmetic table, executed on CPU.
- **Not checked:** I do not have an SM90+ GPU and have **not** executed this on FA3 hardware, so I have
  not confirmed end-to-end that a supported configuration reaches `finalize()` with a non-finite
  `log_sink`. The argument that it can is an inference from the source, not an isolated observation.

Happy to send this as a PR, add a regression case alongside the existing extreme-logits tests, or close
it if the input domain is intended to exclude non-finite sinks — in which case a validation check or a
documented precondition would be the clearer fix.

**Comment by bkryu** [comment]
Hi @headinthebox, thanks for filing the issue.

While the issue is valid, clamp based solutions can cost performance so I'd like to ask whether these issues are detected from fuzz/stress testings or from real workloads. As a library, FlashInfer strives to provide correct and performant support, but there are corner cases where a fix can entail a cost.

Can you help me understand whether the issue arises in a real inference case?

**Comment by headinthebox** [comment]
Thanks — that's the right question, and the honest answer changes what I'm asking for.

**Provenance first: this came from source inspection, not from fuzzing and not from a workload.** I was
reading `variants.py` after PR #4401 and noticed that `AttentionSink::update_m_d` (:39-43) clamps its
exponent differences while `OnlineSoftmaxWithSink::finalize` (:117-125), 75 lines below in the same file,
does not. I filed the asymmetry. I have no production incident behind it.

**And on your actual question, I have to concede the point.** I swept the finalize arithmetic over finite
sinks — both signs, magnitudes from 1e-0 to 1e30, against ordinary rows (`d >= 1`) and against the
fully-masked row (`m = -inf, d = 0`): **364 cases, zero failures, minimum `d_new` observed = 1.0.**

That is not a coincidence, it's structural. With `m_new = max(m, log_sink)`:

- `log_sink <= m` gives `scale = exp2(0) = 1` and `d_new = exp2(log_sink - m) + d >= d >= 1`, since the
  row max contributes `exp2(0) = 1` to `d`.
- `log_sink > m` gives `d_new = exp2(0) + d*exp2(m - log_sink) >= 1`.
- The fully-masked row (`m = -inf`) with any finite sink gives `scale = 0, d_new = 1` — correct output 0.

So for any finite sink, no exponent difference is NaN and `d_new >= 1`. **Both the clamps and the
`d_new > 0` guard are unreachable, and I can't show a real inference case.** Sinks are learned per-head
scalars; they're finite. On this evidence I would not argue for a clamp on a hot path, and your
performance concern is well founded — the 1.8% prefill regression the per-element form cost in #4401 is
a real price against a hazard I can't reach.

Two things I still think are worth a moment, and then I'm happy for this to be closed.

**1. The asymmetry itself is a fact about the code, whichever way it resolves.** The same computation is
written two ways in one file. Either FA2's clamps are paying for something unreachable there too, or
FA3's absence is a latent hole. It'd be worth someone deciding which — a comment on `finalize` saying
"finite sinks only, see #4807" would close it as well as a clamp would, at zero cost.

**2. If a non-finite sink ever does arrive, the failure is silent and total, not a small error.**
`scores_scale(mi)` multiplies every column of the row at `:139` and folds into the LSE at `:126`, so one
NaN takes the entire output row and its LSE. `sinks` is not validated anywhere in `prefill.py` and is
documented as "per-head attention-sink logits, dtype float32" — and `-inf` is the conventional
"contributes nothing" logit, which is a plausible way for a caller to disable one head's sink while
keeping the tensor shape. The FA2 path constructs `-inf` for `log_sink` itself; FA3's constructor
(`:157`) passes `sink[h] * log2e` straight through with no fallback.

**If that's worth guarding, it doesn't have to cost the kernel anything.** A one-time
`torch.isfinite(sinks).all()` in the wrapper is O(num_heads) per call rather than per-row per-tile, and
it turns an undefined input into a clear error at the boundary. I'd be glad to send that as a PR instead,
if you'd prefer it to a kernel change — or to close this if you'd rather treat non-finite sinks as
out-of-contract, which is a perfectly reasonable call.

Sorry for not leading with the reachability analysis; I should have done that sweep before filing.

**Comment by bkryu** [comment]
Thanks @headinthebox,
1. Thanks for pointing out the asymmetry in FA2. The fix in #4401 was tuned to make sure that we do not lose performance meaningfully. Nevertheless, I recognize that before the addressing #4401, I should have asked a question similar to this one. It is possible that the clamping was overkill
2. On the non-finite sink, yes the failure can be silent, but it will be unlikely in a functioning production scenario so would be an acceptable cost to have vs. clamping and potentially paying a cost.

I appreciate you pointing out these vulnerabilities; they are always valuable, even if we do not act on them because these concerns help us make conscious decisions. For the current issue though, it seems like no action is needed so I will close it.
